### ISSUE TITLE
Q: Healing results in a disallowed node warning, am I using schism correctly?

### ISSUE THREAD

**Issue by elvanja** [keathley/schism#3]
Hi,

While playing with this in tests I wanted to create a cluster of 2 nodes (besides main/manager node), execute some actions, heal that cluster and then test if reconciling of data between those nodes worked as expected.

Even though the nodes see each other after healing, I get the `17:00:21.793 [error] ** Connection attempt from disallowed node :"test1@127.0.0.1" **` error.

Here's how test looks (omitted boilerplate):
```elixir
[node1, node2] = LocalCluster.start_nodes("test", 2)

Schism.partition([node1])
Schism.partition([node2])

# do some tests, but for sake of this conversation:
IO.inspect(:rpc.call(node1, Node, :list, []), label: ">> node1 after partition")
IO.inspect(:rpc.call(node2, Node, :list, []), label: ">> node2 after partition")

Schism.heal([node1, node2])

# inspect state after healing
IO.inspect(:rpc.call(node1, Node, :list, []), label: ">> node1 after healing")
IO.inspect(:rpc.call(node2, Node, :list, []), label: ">> node2 after healing")
```

And the output is, correctly:
```
>> node1 after partition: [:"manager@127.0.0.1"]
>> node2 after partition: [:"manager@127.0.0.1"]
>> node1 after healing: [:"manager@127.0.0.1", :"test2@127.0.0.1"]
>> node2 after healing: [:"manager@127.0.0.1", :"test1@127.0.0.1"]
```

It heals correctly. However, that error shows. If I'm not mistaken, it is caused because during healing a node changes it's cookie and tries to connect to other nodes, of which some may still have the random cookie. So I the solution was to simply change the healing code like this:
```elixir
  def heal(nodes) do
    for n <- nodes do
      # Reset the node and ensure we can still talk with it
      true = :rpc.call(n, :erlang, :set_cookie, [n, :erlang.get_cookie()])
      :pong = Node.ping(n)
    end

    for n <- nodes do
      for other_node <- (nodes -- [n]) do
        :rpc.call(n, Node, :connect, [other_node])
      end
    end
  end
```
Basically splits it in two loops, one for setting the cookie, another for connecting the nodes (with a small optimisation that skips connecting to self). I'm OK with turning this into a PR, but am not sure if I'm using it correctly in the first place.

I guess I'm in need of a sanity check 😄

**Comment by keathley** [comment]
Thanks, @elvanja! I think your intuition here is correct. We're attempting to connect nodes before all of their cookies have been set. So this fix seems correct to me. If you'd like to open a PR that would be great 👍

**Comment by elvanja** [comment]
Put something up in #4. It doesn't solve the problem 100% but it is an improvement.

And a note, I actually came across the "flutter" problem described in PR because I used `ExUnit.setup_all` to setup the nodes once, used partitioning in tests and then heal via `ExUnit.setup -> on_exit` to get the nodes back to their original state after each test. This was due to setup time for local cluster, wanted to keep test suite as fast as possible.

One solution would be to just write all tests in one test, but that's actually the same as the idea above, I'd still have the problem of partitioning and healing the same local cluster. Another was to create and drop the local cluster for each test, but that increased test set duration considerably.

Hence all this rant here and related PR to try to optimise it all a bit.

A side note, I used gossip strategy for `:libcluster`, and that caused a lot of problems in tests. Setting up test topology like this helped a lot:
```elixir
config :abbr,
  topologies: [
    local: [
      strategy: Cluster.Strategy.Epmd,
      config: [:hosts, []],
      connect: {:net_kernel, :connect_node, []},
      disconnect: {:erlang, :disconnect_node, []},
      list_nodes: {:erlang, :nodes, [:connected]}
    ]
  ]
```

**Comment by keathley** [comment]
Thanks for the info! I typically use `setup_all` and `on_exit` as well. Testing these sorts of partitions and race-conditions is inherently really messy so I just accepted the noise as part of that mess. But cutting down on it certainly helps.

I've typically used the gossip strategy in libcluster. I'd be curious to know what problems you found. Feel free to add them here but I'll go ahead and close this issue for now.

**Comment by elvanja** [comment]
I found that gossip in tests increases the warnings/connection errors noise noticeably (I suppose that's true for gossip in general 😄), even with #4. The reason I suppose is that gossip strategy doesn't discriminate and tries to connect all the nodes all the time and we run into synchronization issues between schism partitioning and gossipy reconnecting. Maybe it could be made smarter, didn't dive in that direction so can't tell.

Anyway, https://github.com/whitfin/local-cluster spawned test nodes are connected with manager (and between them) regardless of libcluster. So in the end, telling libcluster to not connect anything during tests worked just fine. I went with that inspired by https://github.com/bitwalker/libcluster/issues/120#issuecomment-573803375, and if I understood correctly, when application starts, Empd strategy tries to connect, and doesn't reconnect later. Ideal for tests.

A small nit I had with gossip was that I found it tricky to emulate net split locally (not during tests) simply because it would reconnect immediately. Found a way, you can find it in linked issue above.

No other issues for gossip on production (at least not yet).

BTW, there's a small project I'm playing with, to test out some "have me a few nodes, let's see what happens when I flip the power switch on/off every few seconds" scenarios (aka "distributed beaming"). It's the direct culprit for this issue and related PR, will paste the source link once it's up on GH (given your blessing).

Thank you for the assistance and this library! 😄
